framework/src/executor/mod.rs

```diff
@@ -227,6 +231,27 @@ impl<S: 'static + Storage, DB: 'static + TrieDB, Mapping: 'static + ServiceMappi
         }
     }
 
+    fn call_with_tx_hooks(
+        &self,
+        context: ServiceContext,
+        exec_type: ExecType,
+    ) -> ProtocolResult<ExecResp> {
+        let mut tx_hook_services = vec![];
+        for name in self.service_mapping.list_service_name().into_iter() {
+            let sdk = self.get_sdk(&name)?;
+            let tx_hook_service = self.service_mapping.get_service(name.as_str(), sdk)?;
+            tx_hook_services.push(tx_hook_service);
+        }
+        for tx_hook_service in tx_hook_services.iter_mut() {
+            tx_hook_service.tx_hook_before_(context.clone())?;
+        }
+        let original_res = self.call(context.clone(), exec_type);
```

Comment:
`tx_hook_after` might make different choices depending on whether or not an error occurred in the execution, but for now we can ignore it.

Adding a TODO ensures that it will not be forgotten in the future when it is perfected.